### ISSUE TITLE
Bug: [no-misused-promises] An array of promises passed with spread (...args) syntax is not checked

### ISSUE THREAD

**Issue by Aaron1011** [typescript-eslint/typescript-eslint#5744]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=GYVwdgxgLglg9mABBBBnEBbApgCgHQED6EAhgDZkBGJEA1qgFyICCATqyQJ4A8OAlIgC8APkQA3ODAAmwvkwnTEAb0QBfAFDqyWKMkqohiANr8hogAqs4GGKix5WWVHDJjcUViCx8ANIlMiiJbWtvaOzq7unt4AugDc6ihg6Nj4BBD6fEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function consume(..._callbacks: Array<() => void>): void { }

let cbs = [() => Promise.resolve(true), () => Promise.resolve(true)];
consume(...cbs)
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-misused-promises": "error"
  },
};
```

### Expected Result

I expected that the 4th line should report "Promise returned in function argument where a void return was expected.", just as if I manually called the function with each value from the array (`consume(cbs[0], consume(cbs[1])`)

### Actual Result

There was no error reported on the fourth line.

### Additional Info

Handling of 'rest' parameters is implemented in https://github.com/typescript-eslint/typescript-eslint/pull/5731 . However, support for an explicit 'spread' argument still needs to be added.

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.39.0` |
| `@typescript-eslint/parser` | `5.39.0` |
| `TypeScript` | `4.8.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by kirkwaiblinger** [comment]
I think this may have been fixed. When I click on the playground link, I do see an error as described in the Expected Result.

**Comment by kirkwaiblinger** [comment]
Looking a little deeper, it looks like the test case in the issue is checked because in a scenario like 
```ts
declare const f: (...args: Array<() => void>);
declare argsList: Array<() => Promise<void>>;
f(...argsList)
```
the rule will ask "what is the type of `...args`" and the response is `() => Promise<void>`. See https://github.com/typescript-eslint/typescript-eslint/blob/0ca3a11db0695791f7975cc9fc59e95068768341/packages/eslint-plugin/src/rules/no-misused-promises.ts#L672-L683

So, the test cases
```ts
function consume(..._callbacks: Array<() => void>): void { }
// array
let cbs = [() => Promise.resolve(true), () => Promise.resolve(true)];
consume(...cbs)
```
and
```ts
function consume(..._callbacks: Array<() => void>): void { }
// tuple
const cbs = [() => Promise.resolve(true), () => Promise.resolve(true)] as const;
consume(...cbs)
```
pass just fine.

This does reveal though, the rule isn't too careful about checking unions of call signatures, so something like the following will fail to report, though it would be possible to detect.
```ts
function consume(...maybeCallbacks: Array<(() => void) | number>): void { }
const cbs = [() => Promise.resolve(true), 3] as const;
consume(...cbs)
```

That's not isolated to spread arguments; we can just as easily trick the rule with 
```ts
function consume(...maybeCallbacks: Array<(() => void) | number>): void { }
declare const cb: (() => Promise<boolean>) | 3
consume(cb)
```
However, the use of spread args might be a more compelling case to address it than on its own.

Furthermore, while the example in the issue does pass, there is no test case at present to ensure that it passes.

So, I'd see as actionable
1. Add test case from issue
2. Address unions of potentially misused promises (might be better to file a new issue for that, rather than include it in this one)
3. Remove comments in code that refer to this issue and spread elements being unsupported https://github.com/typescript-eslint/typescript-eslint/blob/0ca3a11db0695791f7975cc9fc59e95068768341/packages/eslint-plugin/src/rules/no-misused-promises.ts#L552-L553

Would maintainers agree with these actions?

**Comment by JoshuaKGoldberg** [comment]
Yup, thanks for the deep dive @kirkwaiblinger! All sounds good to me. 🚀
